ArmPkg/Drivers/ArmGicDxe/GicV5/ArmGicV5Dxe.c

```diff
@@ -0,0 +1,582 @@
+/*++
+
+Copyright (c) 2025, ARM Limited. All rights reserved.
+
+SPDX-License-Identifier: BSD-2-Clause-Patent
+
+Module Name:
+
+  ArmGicV5Dxe.c
+
+Abstract:
+
+  Driver implementing the GICv5 interrupt controller protocol
+
+--*/
+
+#include <Library/IoLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/PcdLib.h>
+#include <Library/TimerLib.h>
+#include <Protocol/Cpu.h>
+#include <Protocol/HardwareInterrupt.h>
+#include <Protocol/HardwareInterrupt2.h>
+
+#include "ArmGicDxe.h"
+#include "ArmGicV5.h"
+
+STATIC UINTN                       mGicIrsConfigFrameBase;
+STATIC UINT32                      SpiRange;
+STATIC UINTN                       mGicNumInterrupts;
+STATIC HARDWARE_INTERRUPT_HANDLER  *mRegisteredInterruptHandlers;
+
+EFI_HARDWARE_INTERRUPT_PROTOCOL  gHardwareInterruptV5Protocol;
+
+/**
+  Wait for IRS SPI writes to complete, and return valid status.
+
+  @retval TRUE   Writes have completed successfully, register status is valid
+  @retval FALSE  Writes have timed out, or register status is invalid.
+**/
+STATIC
+BOOLEAN
+WaitForIrsSpiIdle (
+  )
+{
+  UINT32  Timeout = GICV5_IRS_IDLE_TIMEOUT_MS;
+
+  while (!(MmioRead32 (mGicIrsConfigFrameBase + GICV5_IRS_SPI_STATUSR) & GICV5_IRS_SPI_STATUSR_IDLE) && (Timeout > 0)) {
+    MicroSecondDelay (1);
+    Timeout--;
+  }
+
+  if (!(MmioRead32 (mGicIrsConfigFrameBase + GICV5_IRS_SPI_STATUSR) & GICV5_IRS_SPI_STATUSR_IDLE)) {
+    DEBUG ((DEBUG_ERROR, "WaitForIrsSpiIdle: timed out\n"));
+    return FALSE;
+  }
+
+  if (!(MmioRead32 (mGicIrsConfigFrameBase + GICV5_IRS_SPI_STATUSR) & GICV5_IRS_SPI_STATUSR_V)) {
+    DEBUG ((DEBUG_ERROR, "WaitForIrsSpiIdle: status invalid\n"));
+    return FALSE;
+  }
+
+  return TRUE;
+}
+
+/**
+  Enable interrupt source Source.
+
+  @param This     Instance pointer for this protocol
+  @param Source   Hardware source of the interrupt
+
+  @retval EFI_SUCCESS       Source interrupt enabled.
+  @retval EFI_UNSUPPORTED   Invalid interrupt source number.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+GicV5EnableInterruptSource (
+  IN EFI_HARDWARE_INTERRUPT_PROTOCOL  *This,
+  IN HARDWARE_INTERRUPT_SOURCE        Source
+  )
+{
+  UINT32  InterruptType = Source & GICV5_INTERRUPT_TYPE_MASK;
+  UINT32  InterruptId   = Source & GICV5_INTERRUPT_ID_MASK;
+
+  if (InterruptType == GICV5_INTERRUPT_TYPE_SPI) {
+    ArmGicV5SpiEnable (Source);
+  } else if (InterruptType == GICV5_INTERRUPT_TYPE_PPI) {
+    if (InterruptId >= GICV5_NUM_PPI_INTERRUPTS) {
+      DEBUG ((DEBUG_ERROR, "GicV5EnableInterruptSource: invalid PPI number\n"));
+      ASSERT (FALSE);

```

Comment:
`ASSERT (FALSE)` does not produce a very helpful diagnostic in the DEBUG log. Better to assert on the actual condition being checked here: `ASSERT (InterruptId < GICV5_NUM_PPI_INTERRUPTS)`